api/build.go

```diff
@@ -1262,6 +1262,54 @@ func CancelBuild(c *gin.Context) {
 	}
 
 	// build has been abandoned
-	retErr := fmt.Errorf("unable to find a running build for %s/%d", r.GetFullName(), b.GetNumber())
-	util.HandleError(c, http.StatusInternalServerError, retErr)
+	// update the status in the build table
+	b.SetStatus(constants.StatusCanceled)
+	err = database.FromContext(c).UpdateBuild(b)
+	if err != nil {
+		retErr := fmt.Errorf("unable to update status for build %d: %w", b.Number, err)
+		util.HandleError(c, http.StatusInternalServerError, retErr)
+		return
+	}
+
+	// retrieve the steps for the build from the step table
+	steps := []*library.Step{}
```

Comment:
I see that we're adding logic to handle canceling the `steps` attached to the build.

Thoughts on replicating that logic so we also handle `services` attached to the build?

Reply:
Can't believe that didn't cross my mind. Let me add that quick!

Reply:
Sounds good 👍

Reply:
This function is quite the lengthy one though. Not sure how to refactor it to make this better, but we could consider adding some helper functions for getting all the steps/services for a build. Good with adding the logic here for now